include/deal.II/matrix_free/cuda_fe_evaluation.cuh

```diff
@@ -0,0 +1,385 @@
+     * @p get_value() or @p get_gradient() give useful information.
+     */
+    __device__ void evaluate(const bool evaluate_val,
+                             const bool evaluate_grad);
+
+    /**
+     * This function takes the values and/or gradients that are stored on
+     * quadrature points, tests them by all the basis functions/gradients on
+     * the cell and performs the cell integration. The two function arguments
+     * @p integrate_val and @p integrate_grad are used to enable/disable some
+     * of the values or the gradients.
+     */
+    __device__ void integrate(const bool integrate_val,
+                              const bool integrate_grad);
+
+    /**
+     * Return the value of a finite element function at quadrature point
+     * number @p q_point after a call to @p evalue(true,...).
+     */
+    __device__ value_type get_value(const unsigned int q_point) const;
+
+    /**
+     * Write a value to the field containing the values on quadrature points
+     * with component @p q_point. Access to the same fiels as through @p
+     * get_value(), This specifies the value which is tested by all basis
+     * function on the current cell and integrated over.
+     */
+    __device__ void submit_value(const value_type &val_in,
+                                 const unsigned int q_point);
+
+    /**
+     * Return the gradient of a finite element function at quadrature point
+     * number @p q_point after a call to @p evaluate(...,true).
+     */
+    __device__ gradient_type get_gradient(const unsigned int q_point) const;
+
+    /**
+     * Write a contribution that is tested by the gradient to the field
+     * containing the values on quadrature points with component @p q_point
+     */
+    __device__ void submit_gradient(const gradient_type &grad_in,
+                                    const unsigned int q_point);
+
+    /**
+     * Apply the function @p func on every quadrature point.
+     */
+    template <typename functor>
+    __device__ void apply_quad_point_operations(const functor &func);
+
+  private:
+    unsigned int *local_to_global;
+    unsigned int n_cells;
+    unsigned int padding_length;
+
+    const unsigned int constraint_mask;
+
+    Number *inv_jac;
+    Number *JxW;
+
+    // Internal buffer
+    Number *values;
+    Number *gradients[dim];
+  };
+
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, int n_components_,
+            typename Number>
+  __device__
+  FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
+  FEEvaluation(int cell_id,
+               const data_type *data,
+               SharedData<dim,Number> *shdata)
+    :
+    n_cells(data->n_cells),
+    padding_length(data->padding_length),
+    constraint_mask(data->constraint_mask[cell_id]),
+    values(shdata->values)
+  {
+    local_to_global = data->local_to_global + padding_length*cell_id;
+    inv_jac = data->inv_jacobian + padding_length*cell_id;
```

Comment:
This is because the layout used here is the following:

    (d1*dim+d2)*rowlength*ncells + k*rowlength + q

for cell index `k`, quadrature point index `q`, and jacobian component indices `d1` and `d2`.
Of course, this ordering could have been used instead:

    k*rowlength*dim*dim + (d1*dim+d2)*rowlength + q

But the performance is probably about the same as long as `q` is the innermost index (for a structure-of-array layout in effect).